### PR TITLE
Allow to pass and override default hard coded api version

### PR DESCRIPTION
**Pull request by vbauzysvmware**
This change is driven from VM internal disk PR(https://github.com/vmware/go-vcloud-director/pull/272) and investigation. Also we saw need that in near feature in other cases, as we face more such cases when we need specific version due limitation of using lowest possible API version.

* Moved  field supportedVersions from vCDclient to client
* Updated all references with cange of supportedVersions
* Added request handling methods `WithApiVersion` which allows to pass API version which will be written in request header(override default hard coded one)

We need override only per request base API version. Changing connection API version would create compatibility issues and breakage.

### CODE REVIEW

File: govcd/api.go [hunk 1]
```diff
@@ -125,8 +127,13 @@ func ContainsNotFound(err error) bool {
 	return err != nil && strings.Contains(err.Error(), ErrorEntityNotFound.Error())
 }
 
-// NewRequestWitNotEncodedParams allows passing complex values params that shouldn't be encoded like for queries. e.g. /query?filter=name=foo
+// NewRequestWitNotEncodedParamsWithApi allows passing complex values params that shouldn't be encoded like for queries. e.g. /query?filter=name=foo
```
Comment (lvirbalas) [inline govcd/api.go]: `NewRequestWitNotEncodedParamsWithApi` doesn't seem to match real function name.

File: govcd/api.go [hunk 2]
```diff
@@ -125,8 +127,13 @@ func ContainsNotFound(err error) bool {
...
 func (cli *Client) NewRequestWitNotEncodedParams(params map[string]string, notEncodedParams map[string]string, method string, reqUrl url.URL, body io.Reader) *http.Request {
+	return cli.NewRequestWitNotEncodedParamsWithApiVersion(params, notEncodedParams, method, reqUrl, body, cli.APIVersion)
+}
+
+// NewRequestWitNotEncodedParamsWithApiVersion allows passing complex values params that shouldn't be encoded like for queries. e.g. /query?filter=name=foo
+func (cli *Client) NewRequestWitNotEncodedParamsWithApiVersion(params map[string]string, notEncodedParams map[string]string, method string, reqUrl url.URL, body io.Reader, apiVersion string) *http.Request {
```
Comment (lvirbalas) [inline govcd/api.go]: I would like to see additional comment about how to use `apiVersion` field.

File: govcd/api.go (hunk 1 above)
Comment (vbauzysvmware) [reply govcd/api.go]: updated

File: govcd/api.go (hunk 2 above)
Comment (vbauzysvmware) [reply govcd/api.go]: added

File: govcd/api.go [hunk 3]
```diff
@@ -188,6 +196,12 @@ func (cli *Client) NewRequest(params map[string]string, method string, reqUrl ur
 	return cli.NewRequestWitNotEncodedParams(params, nil, method, reqUrl, body)
 }
 
+// NewRequest creates a new HTTP request and applies necessary auth headers if set.
```
Comment (dataclouder) [inline govcd/api.go]:
```suggestion
// NewRequestWithApiVersion creates a new HTTP request and applies necessary auth headers if set.
```

File: govcd/api.go [hunk 4]
```diff
@@ -127,6 +129,12 @@ func ContainsNotFound(err error) bool {
 
 // NewRequestWitNotEncodedParams allows passing complex values params that shouldn't be encoded like for queries. e.g. /query?filter=name=foo
 func (cli *Client) NewRequestWitNotEncodedParams(params map[string]string, notEncodedParams map[string]string, method string, reqUrl url.URL, body io.Reader) *http.Request {
+	return cli.NewRequestWitNotEncodedParamsWithApiVersion(params, notEncodedParams, method, reqUrl, body, cli.APIVersion)
```
Comment (dataclouder) [inline govcd/api.go]: I'd like to see a few tests for the new functions, to make sure that:
1. They are really using the intended API version.
2. Subsequent requests using the regular `Execute*` methods are not affected.

Comment (vbauzysvmware) [reply govcd/api.go]: Could you share your idea how to test that?

File: govcd/api.go (hunk 3 above)
Comment (vbauzysvmware) [reply govcd/api.go]: udpated

File: govcd/api.go (hunk 4 above)
Comment (dataclouder) [reply govcd/api.go]: Here's a possible way:
1. Add fields to a type that was enhanced in API versions later than the current one.
2. Run a GET that retrieves the type with the enhanced fields
3. Make sure the fields are non-empty
4. Run the same query without the new API version
5. Make sure that the enhanced fields are empty

Comment (Didainius) [reply govcd/api.go]: Two more ideas:
* https://golang.org/pkg/net/http/httptest/
or 
* IIRC when you send a request on a specific version, vCD responds with API version (in headers). You could probably check that.

Comment (vbauzysvmware) [reply govcd/api.go]: first option could be valid with on specific vCD version or I have to find valid for all versions, which is lottery
I'll check possibility with response headers

Comment (vbauzysvmware) [reply govcd/api.go]: Test added

File: govcd/api_test.go [hunk 5]
```diff
@@ -95,3 +100,23 @@ func TestTags(t *testing.T) {
 		showTags()
 	}
 }
+
+// Test_NewRequestWitNotEncodedParamsWithApiVersion verifies that api version override works
+func (vcd *TestVCD) Test_NewRequestWitNotEncodedParamsWithApiVersion(check *C) {
+	fmt.Printf("Running: %s\n", check.TestName())
```
Comment (Didainius) [inline govcd/api_test.go]: Double check the outputs as they are missing new lines. Heres' how it looks.

```
START: api_test.go:105: TestVCD.Test_NewRequestWitNotEncodedParamsWithApiVersion
Running: TestVCD.Test_NewRequestWitNotEncodedParamsWithApiVersion
Test: TestVCD.Test_NewRequestWitNotEncodedParamsWithApiVersion run with api Version: 33.0
PASS: api_test.go:105: TestVCD.Test_NewRequestWitNotEncodedParamsWithApiVersion 0.191s
```

Comment (Didainius) [reply govcd/api_test.go]: Ok. This is fine as it is as planned.

**Pull request merged**